### DISCUSSION TITLE
Preference Modification

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#9880] [General] [answered]
Hi All,
I'm trying to implement a new Preference Menu and so, I've unbind CorePreference and then bind my implementation of CorePreference1 (Same as CorePreference). But, I'm getting "Failed to start the frontend application" Error. It would be great if you could help.

Thanks

![image](https://user-images.githubusercontent.com/80001447/128907558-9609473c-fdff-4b9f-ad0c-8405cf89ff43.png)

The following would be my custom CorePreference.

```ts
import { CorePreferences, createPreferenceProxy, PreferenceContribution, PreferenceProxy, PreferenceSchema, PreferenceService } from '@theia/core/lib/browser';
import { interfaces } from 'inversify';

export const corePreferenceSchema: PreferenceSchema = {
    'type': 'object',
    properties: {
        'workbench.list.openMode': {
            type: 'string',
            enum: [
                'singleClick',
                'doubleClick'
            ],
            default: 'singleClick',
            description: 'Controls how to open items in trees using the mouse.'
        },
        'workbench.editor.highlightModifiedTabs': {
            'type': 'boolean',
            'description': 'Controls whether a top border is drawn on modified (dirty) editor tabs or not.',
            'default': false
        },
        'workbench.editor.closeOnFileDelete': {
            'type': 'boolean',
            // eslint-disable-next-line max-len
            'description': 'Controls whether editors showing a file that was opened during the session should close automatically when getting deleted or renamed by some other process. Disabling this will keep the editor open  on such an event. Note that deleting from within the application will always close the editor and that dirty files will never close to preserve your data.',
            'default': true
        },
        'application.confirmExit': {
            type: 'string',
            enum: [
                'never',
                'ifRequired',
                'always',
            ],
            default: 'ifRequired',
            description: 'When to confirm before closing the application window.',
        },
        'workbench.silentNotifications': {
            type: 'boolean',
            default: false,
            description: 'Controls whether to suppress notification popups.'
        },
        'workbench.tree.renderIndentGuides': {
            type: 'string',
            enum: ['onHover', 'none', 'always'],
            default: 'onHover',
            description: 'Controls whether the tree should render indent guides.'
        },
        'keyboard.dispatch': {
            type: 'string',
            enum: [
                'code',
                'keyCode',
            ],
            default: 'code',
            description: 'Whether to interpret keypresses by the `code` of the physical key, or by the `keyCode` provided by the OS.'
        },
    }
};

export interface CoreConfiguration {
    'application.confirmExit': 'never' | 'ifRequired' | 'always';
    'keyboard.dispatch': 'code' | 'keyCode';
    'workbench.list.openMode': 'singleClick' | 'doubleClick';
    'workbench.commandPalette.history': number;
    'workbench.editor.highlightModifiedTabs': boolean;
    'workbench.editor.closeOnFileDelete': boolean;
    'workbench.colorTheme': string;
    'workbench.iconTheme': string | null;
    'workbench.silentNotifications': boolean;
    'files.encoding': string
    'workbench.tree.renderIndentGuides': 'onHover' | 'none' | 'always';
}

export const CorePreferences1 = Symbol('CorePreferences');
export type CorePreferences1 = PreferenceProxy<CoreConfiguration>;

export function createCorePreferences(preferences: PreferenceService): CorePreferences {
    return createPreferenceProxy(preferences, corePreferenceSchema);
}

export function bindCorePreferences1(bind: interfaces.Bind): void {
    bind(CorePreferences1).toDynamicValue(ctx => {
        const preferences = ctx.container.get<PreferenceService>(PreferenceService);
        return createCorePreferences(preferences);
    }).inSingletonScope();
    bind(PreferenceContribution).toConstantValue({ schema: corePreferenceSchema });
}
````

![image](https://user-images.githubusercontent.com/80001447/128907646-16f16326-58fd-4ce0-bf92-2c440baa488e.png)

**Comment by vince-fugnitto** [comment]
@maheshrajrp I'm not sure I follow as to why you would want to `rebind` the core-preferences like you've done in order to create a custom menu, but https://github.com/eclipse-theia/theia/discussions/9880#discussioncomment-1155511 should be the resolution.

**Reply by maheshrajrp** [reply]
@vince-fugnitto Thanks for the reply. 

My aim is to remove most of the preference items and add only those that are absolutely necessary. The code which I've attached is a sample code. I was exploring my options when I stumbled upon this issue.

**Reply by vince-fugnitto** [reply]
I believe such a change is too complex, and will not scale well when building an application that has many `theia` extensions, and plugins that also contribute preferences, but if it's your use-case then good luck 👍 I am also not sure what the behavior will be after the change, since there is code that references preferences directly.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Yeah, I'm able to understand. But, what would be your suggestion if I must remove a few preference items, say I wish to remove Editor Font Size and force the user to use the default values and make Editor Font Size hidden from the preferences menu.

Thanks

**Comment by paul-marechal** [comment]
```ts
export const CorePreferences1 = Symbol('CorePreferences');
```

You are creating a new `Symbol` which is like a unique reference. Services expect `CorePreferences` precisely to be bound.

**Reply by paul-marechal** [reply]
The `No matching bindings for serviceIdentifier: Symbol(CorePreferences)` error hints that this is likely the issue.

**Reply by maheshrajrp** [reply]
Hi @paul-marechal ,
Thanks for the reply. Yes, I'm able to understand the issue. However, is there any workaround for this issue ? I really intend to remove the preferences and add my own.

Thanks in advance

**Reply by vince-fugnitto** [reply]
Paul already mentions the fix, you are creating a new `symbol` rather than re-using the existing one.

```ts
export function bindCorePreferences1(bind: interfaces.Bind): void {
    bind(CorePreferences).toDynamicValue(ctx => {
        const preferences = ctx.container.get<PreferenceService>(PreferenceService);
        return createCorePreferences(preferences);
    }).inSingletonScope();
    bind(PreferenceContribution).toConstantValue({ schema: corePreferenceSchema });
}
```

**Reply by paul-marechal** [reply]
Make sure to update the `PreferenceContribution` to remove the `corePreferenceSchema`. If you try to to `unbind(PreferenceContribution)` you might end up unbinding all contributions registered so far. Maybe look at contribution filters.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto and @paul-marechal ,

I tried this too. It still doesn't work. The code is below. Kindly assist me.
Thanks

```
import { CorePreferences, FrontendApplication, FrontendApplicationContribution, PreferenceService } from '@theia/core/lib/browser';
import { CorePreferences1 } from './sample-preferences/sample-preferences-settings';
    @inject(CorePreferences)
    protected readonly preferences: CorePreferences1; 
    ..... ...... .... (Some Lines Here) 
    async configure(app: FrontendApplication): Promise<void> {
        this.updateIconSizeFromPreference('sample.sample.iconSize');
        this.preferences.onPreferenceChanged(e => {
            if (e.preferenceName === 'sample.sample.iconSize') {
                this.updateIconSizeFromPreference(e.preferenceName);
            }
        });
    }
```

common-frontend.tsx
```

    unbind(CorePreferences)
    bindCorePreferences1(bind);

```
```
import { CorePreferences, createPreferenceProxy, PreferenceContribution, PreferenceProxy, PreferenceSchema, PreferenceService } from '@theia/core/lib/browser';
import { interfaces } from 'inversify';


export const corePreferenceSchema: PreferenceSchema = {
    'type': 'object',
    properties: {
        'sample.sample.iconSize': {
            type: 'string',
            enum: [
                'small',
                'medium',
                'large'
            ],
            default: 'small',
            description: 'Select the icon size.'
        },
    }
};

export interface CoreConfiguration {
    'sample.sample.iconSize': 'small' | 'medium' | 'large';
}
export type CorePreferences1 = PreferenceProxy<CoreConfiguration>;

export function createCorePreferences(preferences: PreferenceService): CorePreferences1 {
    return createPreferenceProxy(preferences, corePreferenceSchema);
}

export function bindCorePreferences1(bind: interfaces.Bind): void {
    bind(CorePreferences).toDynamicValue(ctx => {
        const preferences = ctx.container.get<PreferenceService>(PreferenceService);
        return createCorePreferences(preferences);
    }).inSingletonScope();
    bind(PreferenceContribution).toConstantValue({ schema: corePreferenceSchema });
}

```

**Comment by maheshrajrp** [comment]
Hi @vince-fugnitto ,

Thanks for the fix. I was hoping to try this one out. However, was unable to locate "WorkspacePreferencesSchemaProvider" which is mentioned in https://github.com/eclipse-theia/theia/commit/c5a7d712d21a7f89eb03c03244ac1b91f38d02a6
 . Can you provide assistance regarding that. 

Thanks

**Reply by vince-fugnitto** [reply]
> Hi @vince-fugnitto ,
> 
> Thanks for the fix. I was hoping to try this one out. However, was unable to locate "WorkspacePreferencesSchemaProvider" which is mentioned in [c5a7d71](https://github.com/eclipse-theia/theia/commit/c5a7d712d21a7f89eb03c03244ac1b91f38d02a6)
> . Can you provide assistance regarding that.
> 
> Thanks

@maheshrajrp its only an example on how to rebind the schema, originally part of the pull-request as a test commit. You can inspire yourself from what it does to rebind schemas as necessary for your use-case in your custom extension 👍

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto, 
I'm still having trouble understanding it. Can you provide any documentation or examples ?

Thanks.

**Reply by vince-fugnitto** [reply]
> Can you provide any documentation or examples ?

Given you have a [custom extension](https://theia-ide.org/docs/authoring_extensions/), the feature allows you to more easily `rebind` the preference-schema (using [inversify](https://github.com/inversify/InversifyJS) like the rest of the framework) to suit your needs.

Thanks to the changes, we are now able to more easily `rebind` the schema like so (ex: `WorkspacePreferenceContribution`):

```ts
export default new ContainerModule((bind, unbind, isBound, rebind) => {
    rebind(WorkspacePreferenceContribution).toConstantValue({ schema: getSchema() });
});

function getSchema(): PreferenceSchema {
    const schema = workspacePreferenceSchema;
    delete schema.properties['workspace.preserveWindow'];
    return schema;
}
```

As a test this change will remove the `workspace.preserveWindow` preference from the original schema.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,

Thanks for the example, I'm able to get it working now. !!

**Comment by vince-fugnitto** [comment, marked as answer]
In case others are looking for an example:
- https://github.com/eclipse-theia/theia/discussions/9880#discussioncomment-1341842
